### DISCUSSION TITLE
Migrating from `did:plc` to `did:web`

### DISCUSSION THREAD

**Discussion by TCFox** [bluesky-social/atproto#2705] [How to do it?] [answered]
Hi folks!

I've signed up for Bluesky a while back in the past, but am really interested in the idea of hosting my identity in a more decentralised manner (largely for geek-cred than anything else). This may or may not also involve running my own Bluesky PDS for personal use.

While migrating to a new PDS is a challenging but supported task, is it also possible to migrate my current `did:plc` identity to a `did:web` identity either separately or at the same time (while keeping my handle the same in all cases)?

My current understanding is this is likely impossible, given it's _the_ identifier for my account, and other relationships (such as followers) will suddenly lose track of me. But I'm unsure if those relationships actually track my handle rather than my `did` identifier, or if there is some sort of "pointer" mechanism built into the `did:plc` document specification that could facilitate any relationships between data that still uses my old `did:plc` identifier.

If it's _definitely not_ possible, what are the consequences of moving my handle from a `did:plc` identity to a different `did:web` identifier? Is there any way I can limit the amount of "breakage" while moving between these two identities (particularly in the context of Bluesky itself)?

I'm keen to hear what might be possible!

**Comment by DavidBuchanan314** [comment, marked as answer]
>  My current understanding is this is likely impossible

Yup, that's pretty much it. You can keep your *handle* and point it at your new DID, but all in-protocol references are based on the DID. You will lose your followers, for example (because they're "following" your DID, not your handle).

You could in theory re-import your old posts, but they'd lose any existing replies/likes/etc., and any threads would be broken.

I might do a similar migration myself at some point, and if I do, here's what I'll do:

1. Migrate the `did:plc` account to my own PDS.
2. Create the new `did:web` account, also on my own PDS.
3. Change the handle on the `did:plc` account to a "legacy." subdomain or something, and move the original handle to the new account.
4. Set up automated ~daily posting on the `did:plc` account to say "hey I've moved to [did:web account], please follow me there instead"

It's useful to keep the `did:plc` account around, both as an archive of old posts (preventing dead links), and as a way to direct people towards the `did:web` account. Notably, I don't plan on trying to import old posts - seems like more trouble than it's worth.

Things like follow*s*, blocks, custom feeds, mutes, preferences could be migrated fairly seamlessly though - I wonder if anyone's made tooling for that.

**Reply by TCFox** [reply]
Couldn't have asked for a better person to answer this (really enjoyed your "Malleable Deputy" writeup!)

Was hoping that there might be some sort of "pointer" mechanism within `did:plc`, but the complexity around that is probably far too difficult to be worth it, particularly updating the protocol and modifying every client to know what to do when the pointer is encountered. I think you're right, I should cut my losses and move across to `plc:web` while my account is still "new". Good idea with the "legacy" handler though!

**Reply by snarfed** [reply]
Agreed, @DavidBuchanan314 is awesome!

The goal with DIDs as identities is that they kind of _are_ a pointer mechanism. They're not tied to any specific handle, PDS, or key(s), so you can switch all of those at will.

`did:plc` has the added benefit of not being tied to control over any domain, which is a nice safeguard, but right now the tradeoff is that it depends on Bluesky team's centralized PLC directory. 🤷

**Reply by bnewbold** [reply]
Yup, this is basically the situation: DIDs are the long-term permanent identifier in the system. Having a layer of indirection where the "permanent" identifier can be migrated/updated/redirected would be a lot of complexity and result in inconsistent behavior. Conceptually the whole point of DIDs is that they *don't* change for the lifetime of the account, so they can be baked in to content-addressed records (etc). IMO the experience of "best effort" redirects other distributed identifier systems has been poor.

Do keep in mind that your `did:web` will only work as long as you maintain the DNS registration and resolution mechanism! That might seem like a "solved problem" in the short term (next few years), but can be hard to maintain in the long run, even for motivated and well-resourced folks. The PLC system is at least potentially more permanent and "recoverable" as infrastructure.

Everybody has their own threat-models and priorities/interests though! We do support `did:web` for a reason and intend to do so indefinitely.

**Reply by apatrida** [reply]
A better did:web is did:webvh to allow moving across domains and tracking history, originally as did:tdw it wasn't a stable identifier, but is the SCID now stable enough to use as the DID within AT protocol?  Seems like they have now (or are in the process) of adding things like witness servers and federated servers for taking a role similar to PLC server, yet without the fault of PLC server that it could be used as a point of policy control.

https://identity.foundation/didwebvh/next/

**Reply by str4d** [reply]
The problem with did:webvh is that its DIDs are not long-term permanent identifiers:
> The did:webvh method-specific identifier contains both the [self-certifying identifier](https://identity.foundation/didwebvh/next/#term:self-certifying-identifier) (SCID) for the DID, and a fully qualified domain name (with an optional path) that is secured by a TLS/SSL certificate.

did:webvh permits the domain name part of the DID to change, which means that the DID can change, which means that it cannot be a long-term permanent identifier. If used within the AT Protocol, any time the domain changed, the rest of AT Proto would consider it to be a new identity, rather than the same identity at a new location. As @bnewbold notes above, this is a trade-off that has been chosen here: restrict AT Proto to the subset of DIDs that are stable for the lifetime of the account, in exchange for complexity reduction and consistent behaviour.

**Reply by str4d** [reply]
Also, if you squint a bit, the SCID from did:webvh is basically a DID PLC. There are a few different design decisions they've made here and there (e.g. rotation keys) that take different trade-offs in the design space, but the SCIDs end up having roughly the same properties as DID PLC. At which point, the major thing that did:webvh provides over DID PLC is the ability for the user to control where their DID Document is canonically fetched from.

**Reply by apatrida** [reply]
@str4d read this a bit differently, the DID is in the format of:

`did:webvh:{SCID}:{host.domain}`

If you take the SCID or the prefix, you have your stable identifier.  Only the tail changes.  And you can use either part of the identifier to resolve to the DID document.

So why not just use the `did:webvh:{SCID}` prefix?  This solves the issue stated by @bnewbold 

Now unsquinting I see the SCID is not PLC, why?

* not one central authority, but distributed (a goal of AT protocol is just that, no?)
* witnesses to changes preventing bad actors (including BlueSky or anyone else can one day become a bad actor from someone's perspective)
* full verifiable history of DID changes

> At which point, the major thing that did:webvh provides over DID PLC is the ability for the user to control where their DID Document is canonically fetched from.

As I just mentioned, this is not the ONLY major thing, and it is a major thing.  Why support `did:web` and risk loss of domain control, yet AT Protocol does.  Why support PLC and have one central authority and risk control of your DID, yet AT Protocol does.  

It is a very BIG thing to resolve this, and `did:webvh` using the identifier up to the SCID prefix fixes this no?

**Reply by apatrida** [reply]
The reason I bring this up, is as we look at support for AT Protocol, the huge push back is that the DID model is fragile and can be used for policy control by either a domain registrar or by BlueSky or by accident (loss of domain).  And therefore some people might not want to invest in the work to later be blacklisted from the distributed system that has centralized controls.

**Reply by apatrida** [reply]
is did:dht also not a stable identifier?

https://did-dht.com/

**Reply by apatrida** [reply]
Another note on `did:webvh` in that I think it is going to be an acceptable identifier for Self-Issued OpenID Provider which is a pretty interesting topic as well.

**Reply by snarfed** [reply]
> So why not just use the `did:webvh:{SCID}` prefix? This solves the issue stated by @bnewbold

The short answer is, because the AT Protocol protocol is currently specified to use the whole DID. It doesn't (yet) have a way to use "transforms" like this per DID method.

The protocol could change to allow that! It seems unlikely, but not impossible. But it doesn't work that way now, and so far we have no indication that it's under consideration for the future.

**Reply by swcurran** [reply]
The [did:webvh implementers guide](https://didwebvh.info/latest/implementers-guide/did-portability/) outlines our approach to DID portability, which aligns closely with @bnewbold's comment. We consider the prefix to be the stable identifier, while the inclusion of the domain/path enables straightforward discoverability. For DIDs that never relocate, this works seamlessly — and `did:webvh` allows for a DID Controller to prevent ever porting their DID. However, for DIDs that do move, our portability design retains the prefix and the verifiable history of the DID, reflecting the decentralized trust model of the Web.

The portability approach was heavily influenced by use cases like Bluesky’s. In this model, a service provider (e.g., Bluesky) could host the DID and provide services like social media, email, chat, or file hosting, utilizing the user’s keys published in the DID. If the user decides to migrate to a new service, they could carry their DID's SCID and history to the new location. This ensures that verifiers can trust that the same entity controls the DID post-move.

This contrasts with the current scenario where users must inform their contacts of new service addresses (e.g., “Hey everyone, I’ve moved from Gmail to Yahoo and here is my new email address.  Really, it's me -- honest!!!”). With did:webvh, the move would be cryptographically verifiable: verifiers could check SCID/log entries to confirm the change, maintaining continuity and trust. Ideally, a service provider like Bluesky could implement resolution redirection for the “old” DID to its new location when a did:webvh moves, enhancing user experience and trustworthiness. extending the lifetime of the original DID.

We also borrowed features from did:plc to enhance usability and security, such as allowing multiple keys to authorize updates. This enables a trusted service provider to assist in managing the DID when necessary. Additionally, our witnessing mechanism could support confirm-before-update features, complementing the design philosophy of did:plc.

Happy to collaborate / discuss further.

**Reply by bnewbold** [reply]
From the standpoint of atproto:

- the DID is a permanent account identifier. accounts are one-to-one with an exact string. the string can not change, in part or in whole
- every server on the internet should be able to rapidly resolve a DID, and to some degree everybody should see the same result. caching is allowed, but should not be required. that is, a brand new server just installed today should have all the same info a 5-year-old server. servers are not expected "store" identity information (aka, should not have lookup or mapping tables internally)
- the DID method needs to describe how to resolve any DID with the method prefix

`did:webvh` is interesting. If it got a lot of independent adoption, or full standardization by the W3C, maybe we could at least partially support it. I think the way that would work is that it would be basically `did:web` with an oplog, but migration would be not supported: a new domain would mean an entirely new DID from the perspective of atproto, and things like social graph (following) and post threads would not be migrated (you'd be starting over with a new account). But this is still arguably better and more secure than `did:web`: there would not be the confusion of a `did:web` account changing hands and suddenly you are "following" an account controlled by a different party.

`did:dht` is interesting and maybe could work, for atproto, but there have been many DHT projects in the history of decentralized computing, and only a small handful have really worked out. AKA, been resilient to abuse, attacks, neglect, etc, over years and decades. magnet links are cool but kind of the exception to the rule. I think if the W3C standardized `did:dht` it could be a real contender, but somebody needs to carry the flag and demonstrate it working "in real life" with millions of users, and I suspect that will be difficult, and I would not expect Bluesky to lead that charge (we already have our own pet DID method to maintain and support).

**Reply by erlend-sh** [reply]
> did:dht is interesting and maybe could work, for atproto

There’s some very interesting work being done with DHT by the folks (@SeverinAlexB & @Nuhvi) at [pubky](https://github.com/pubky):

- [pkarr](https://github.com/pubky/pkarr) — content addressing and retrieval via DHT.
- [pkdns](https://github.com/pubky/pkdns) — DNS server resolving records hosted on the Mainline DHT.

**Reply by erlend-sh** [reply]
pkarr was being used as a fundamental building block in `did:dht`, which unfortunately appears to be [discontinued](https://tbd.website) (afaik this was the org driving it). Even so, @decentralgabe may appreciate the renewed discussion here.

> migration [in `did:webvh`]would be not supported: a new domain would mean an entirely new DID from the perspective of atproto, and things like social graph (following) and post threads would not be migrated (you'd be starting over with a new account).

Doing some of my signature unqualified hand waving, I wonder if pkarr’s ‘sovereign TLDs’ could be used in `did:webvh` as a permanent domain identifier to work around the lossyness of domain migrations.

**Reply by thereisnogabe** [reply]
`did:dht` uses the same underlying tech as pkarr, they had the same roots. however, I believe the overhead of DNS packet-style serialization is inefficient, and would like to see `did:dht` evolve to use dCBOR. @christophera is interested in this too and might have some thoughts here. `did:dht` has been moved from its prior home (TBD) to the [Decentralized Identity Foundation](https://github.com/decentralized-identity/did-dht) which is quite a friendly org for progressing the project.

I'm happy to help however I can.

**Reply by bumblefudge** [reply]
note that dCBOR is not DAG-CBOR.  two unrelated ways of making CBOR deterministic enough for signing and hashing and whatnot. 

https://datatracker.ietf.org/doc/draft-mcnally-deterministic-cbor/

https://datatracker.ietf.org/doc/draft-caballero-cbor-cborc42/

in terms of prior art based on other DHTs than the bittorrent mainnet, it's worth mentioning that very many DID methods have been proposed, prototyped, and in some cases taken all the way to prod based on IPFS:

* did:dht originally started as a v2 of [IPFS-based] sidetree, which used IPFS-compatible (and thus portable) oplogs of each update to a DID. the most well-known is did:ion, but did:element also made it to prod in many cases and exists as a kind of underdog in the "what's the best ethereum-based DID method" sidequest.
* did:ipid is just IPNS reframed as a DID method, never made it to prod AFAIK but kind of the IPFS equivalent of did:dht. basically a free-form publication method where any controller using any IPNS-supported keytype can self-publish/self-pin their DID docs on IPFS mainnet/amino.
* there have been a number of others.  the most recent and interesting to look at would probably be [did:svdx](https://ggreve.medium.com/a-future-for-self-sovereign-identity-c237caa5e46f) and [did:ipfs (possibly renamed since?)](https://github.com/sid030sid/did-ipfs-service-provider?tab=readme-ov-file#didipfs-service-provider)

I think it would be helpful to distinguish between four very orthogonal topics, each of which should probably be a separate line of research/negotiation:

- which DID methods can BlueSky the app _use_ for anything (to identify services and/or account-holders/actors in the system)?
- which DID URIs can be safely (or usefully/meaningfully) added to `alsoKnownAs`?
- which key types (or other metadata) can be added as `verificationMethod`s without blueSky barfing/throwing/rejecting the user/DID doc?
- which service types can be added to the `services` array without blueSky barfing/throwing/rejecting the user/DID doc?

At the moment, BSPBC is the only person who can provide any answer on any of the 4; in theory, did:plc's method and an atproto-wide profile/validation rules for those DIDs could some day be answered by some tbd multistakeholder governance.

**Reply by erlend-sh** [reply]
@bnewbold 

> the DID is a permanent account identifier. accounts are one-to-one with an exact string. the string can not change, in part or in whole

Why is that permanency so important; what’s the affordance?

To me it seems much more important to have a permanent *record* than any singularly permanent ID-key, that way most forms of key-changes can be solved with a redirect in the permanent record of redirects as The Key changes from one to the next. That record itself becomes each person’s most unique digital fingerprint.

**Reply by snarfed** [reply]
Interesting idea!

One main value of permanent DIDs is supporting global/large scale aggregation and indexing in appviews. With permanent DIDs, you can find all of an account's followers, follows, blocks, etc by just querying for records that reference its DID. You couldn't do that if its DID changed over time.

Having said that, an appview could track accounts' DID changes over time, and query for all of them at the same time. Definitely less efficient/scalable, and you'd probably need to limit the number of times an account can change its DID, but it is technically possible.

Making this work for PDSes would be more difficult. Repos are identified by DID, at:// URIs embed the DID in them, and MSTs are optimized to look up records by those URIs. It's not clear how to modify the current design to support lookups by an arbitrary number of historical DIDs.

did:web also makes this difficult, since there's no global feed of did:web changes that an appview or relay could follow to detect all changes.

Similarly, did:web redirects would depend on the old domain and server serving the redirected DID doc indefinitely, which often doesn't happen, at we've seen in the fediverse.

**Comment by swcurran** [comment]
Thanks for the feedback @bnewbold — interesting.  In creating `did:webvh`, we need web resolution as a core feature (it’s in the name, after all…), so having the location in the DID is pretty important. Further, the web changes over time, and so allowing portability is likewise deemed necessary (although it is not required — a `did:webvh` DID can state it is “not moveable”).  Interesting to think about if we could make the DID end at the SCID (globally unique and long-lasting) and use some sort of DID URL mechanism to allow for the location of the DID.

I do think being able to move a DID as done with `did:webvh` could be a powerful enabler of “credible exit” — being able to take your unique identifier (SCID) and your DID history of keys and key rotations when you move to another service.

**Reply by bumblefudge** [reply]
i would argue that did:plc IS already web resolveable... it's just resolved from the DID PLC directory instead of from the PDS itself, even though [that step could be skipped by querying the PDS for the DID document on a known or even `.well-known` route](https://docs.bsky.app/docs/advanced-guides/entryway#virtual-pds).  so while did:webvh encodes the domain and the route to GET a did-doc in the DID URI, a few slight tweaks to the PDS API would make a did:webvh deterministically derivable from each did:plc's `service[#atproto_pds].serviceEndpoint]` value. This way, each did doc that wants to be so dual-use could be resolveable by a `did:plc` OR a `did:webvh` (or a `did:web` actually!).

**Reply by bumblefudge** [reply]
as for the portability across PDSs, it is already possible to update a DID to change it's PDS root-of-trust (i.e. its `service[#atproto_pds].serviceEndpoint]` value), IF AND ONLY IF the `alsoKnownAs` also changes in the same update (i.e., if you move off a bsky PDS, you also need to rotate your "atproto handle" from *.bsky.social to something rooted in the same domain apex as your PDS).  if another tweak to the PDS API also obligated the new home to host your entire "oplog" and DID doc history going back to inception (to allow validation of the genesis document against the CID in the did:plc URI), this would close the delta and make `did:plc` a variant/alternate expression of `did:webvh`.

**Reply by bumblefudge** [reply]
maybe i'm missing a corner-case here, but in this universe, each did:plc that a given PDS hosts is a valid did:webvh, but not every valid did:webvh needs to create a did:plc or be intelligible to ATProto.

**Reply by snarfed** [reply]
> it is already possible to update a DID to change it's PDS root-of-trust (i.e. its service[#atproto_pds].serviceEndpoint] value), IF AND ONLY IF the alsoKnownAs also changes in the same update

Hmm, where did you hear this? I get the idea, but isn't actually true. Did you see it in https://github.com/did-method-plc/did-method-plc or one of the atproto.com specs? I don't remember it there.

In any case, I've written (and exercised) a service that changes did:plc service endpoints without also changing their `alsoKnownAs`. https://web.plc.directory/did/did:plc:5zspv27pk4iqtrl2ql2nykjh is an example of a DID doc with a bsky.social handle on a non-bsky.social PDS.

In general, handles and PDSes aren't coupled that tightly, which is good! If you have a custom domain handle, you shouldn't have to change it when you migrate PDSes.

**Reply by bumblefudge** [reply]
Oh, neat, I didn't know you could keep your .bsky handle when you migrated off bsky servers! My natural intelligence must have hallucinated that one. That makes it even easier to make dual-method accounts/documents 😏
